src/BenchmarkDotNet/Engines/Engine.cs

```diff
@@ -214,31 +215,49 @@ private ClockSpan Measure(Action<long> action, long invokeCount)
 
         private (GcStats, ThreadingStats, double) GetExtraStats(IterationData data)
         {
-            // we enable monitoring after main target run, for this single iteration which is executed at the end
-            // so even if we enable AppDomain monitoring in separate process
-            // it does not matter, because we have already obtained the results!
-            EnableMonitoring();
+            // Warm up the measurement functions before starting the actual measurement.
+            DeadCodeEliminationHelper.KeepAliveWithoutBoxing(GcStats.ReadInitial());
+            DeadCodeEliminationHelper.KeepAliveWithoutBoxing(GcStats.ReadFinal());
 
             IterationSetupAction(); // we run iteration setup first, so even if it allocates, it is not included in the results
 
             var initialThreadingStats = ThreadingStats.ReadInitial(); // this method might allocate
             var exceptionsStats = new ExceptionsStats(); // allocates
             exceptionsStats.StartListening(); // this method might allocate
-            var initialGcStats = GcStats.ReadInitial();
 
-            WorkloadAction(data.InvokeCount / data.UnrollFactor);
+            if (RuntimeInformation.IsNetCore && Environment.Version.Major is >= 3 and <= 6 && RuntimeInformation.IsTieredJitEnabled)
+            {
+                // #1542
+                // We put the current thread to sleep so tiered jit can kick in, compile its stuff,
+                // and NOT allocate anything on the background thread when we are measuring allocations.
+                // This is only an issue on netcoreapp3.0 to net6.0. Tiered jit allocations were fixed in net7.0,
```

Comment:
> Tiered jit allocations were fixed in net7.0,

We might have reduced the tiered JIT allocations in .NET 7, but they can still happen. There is nothing fundamental preventing the tiered JIT thread from allocating.

Reply:
I suppose that's true. Do you suggest we always put that thread sleep? I don't think it will do much. The thread sleep isn't even a guarantee on the older runtimes, it just so happens to make it more stable, and it's already much more stable in net7+.

Reply:
My comment was primarily about the comment being false.

This approach taken by this is always going to be unreliable. I do not have an opinion whether the sleep should be there for all runtimes or not.  It is about how much unreliability you are willing to tolerate.